### PR TITLE
federation: replacing string credentials field by secretRef

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/kubernetes/kubernetes/issues/25761

Replaced the string Credentials field in ClusterSpec by secretRef as discussed in https://github.com/kubernetes/kubernetes/issues/25761.
Also updated the clusterController to use this new secretRef field while creating a client to talk to a k8s cluster.

cc @lavalamp @kubernetes/sig-cluster-federation

### CODE REVIEW

File: federation/apis/federation/types.go [hunk 1]
```diff
@@ -37,10 +37,11 @@ type ClusterSpec struct {
 	// Clients can use the appropriate server address as per the CIDR that they match.
 	// In case of multiple matches, clients should use the longest matching CIDR.
 	ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" patchStrategy:"merge" patchMergeKey:"clientCIDR"`
-	// the type (e.g. bearer token, client certificate etc) and data of the credential used to access cluster.
-	// It’s used for system routines (not behalf of users)
-	// TODO: string may not enough, https://github.com/kubernetes/kubernetes/pull/23847#discussion_r59301275
-	Credential string `json:"credential,omitempty"`
+	// Name of the secret containing kubeconfig to access this cluster.
+	// The secret is read from the kubernetes cluster that is hosting federation control plane.
+	// Admin needs to ensure that the required secret exists. Secret should be in the same namespace where federation control plane is hoster and it should have kubeconfig in its data with key "kubeconfig".
```
Comment (madhusudancs) [inline federation/apis/federation/types.go]: "... federation control plane is **hoste~~r~~d**"

File: federation/apis/federation/v1alpha1/types.go [hunk 2]
```diff
@@ -37,10 +37,12 @@ type ClusterSpec struct {
 	// Clients can use the appropriate server address as per the CIDR that they match.
 	// In case of multiple matches, clients should use the longest matching CIDR.
 	ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs" patchStrategy:"merge" patchMergeKey:"clientCIDR" protobuf:"bytes,1,rep,name=serverAddressByClientCIDRs"`
-	// the type (e.g. bearer token, client certificate etc) and data of the credential used to access cluster.
-	// It’s used for system routines (not behalf of users)
-	// TODO: string may not enough, https://github.com/kubernetes/kubernetes/pull/23847#discussion_r59301275
-	Credential string `json:"credential,omitempty" protobuf:"bytes,2,opt,name=credential"`
+	// Name of the secret containing kubeconfig to access this cluster.
+	// The secret is read from the kubernetes cluster that is hosting federation control plane.
+	// Admin needs to ensure that the required secret exists. Secret should be in the same namespace where federation control plane is hoster and it should have kubeconfig in its data with key "kubeconfig".
```
Comment (madhusudancs) [inline federation/apis/federation/v1alpha1/types.go]: Same comment as above.

File: federation/apis/federation/v1alpha1/types.go [hunk 3]
```diff
@@ -37,10 +37,12 @@ type ClusterSpec struct {
...
+	// This will later be changed to a reference to secret in federation control plane when the federation control plane supports secrets.
+	//		SecretRef *v1.LocalObjectReference `json:"secretRef" protobuf:"bytes,2,opt,name=secretRef,casttype=k8s.io/kubernetes/pkg/api/v1.LocalObjectReference"`
```
Comment (madhusudancs) [inline federation/apis/federation/v1alpha1/types.go]: Is this commented code necessary?

File: pkg/client/unversioned/clientcmd/client_config.go [hunk 4]
```diff
@@ -399,3 +399,11 @@ func BuildConfigFromFlags(masterUrl, kubeconfigPath string) (*restclient.Config,
 		&ClientConfigLoadingRules{ExplicitPath: kubeconfigPath},
 		&ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: masterUrl}}).ClientConfig()
 }
+
+// BuildConfigFromKubeconfigGetter is a helper function that builds configs from a master
+// url and a kubeconfigGetter.
+func BuildConfigFromKubeconfigGetter(masterUrl string, kubeconfigGetter KubeconfigGetter) (*restclient.Config, error) {
+	return NewNonInteractiveDeferredLoadingClientConfig(
+		&ClientConfigGetter{kubeconfigGetter: kubeconfigGetter},
+		&ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: masterUrl}}).ClientConfig()
```
Comment (madhusudancs) [inline pkg/client/unversioned/clientcmd/client_config.go]: There are too many things happening in this one statement, could you please reorganize this a bit. It took a bit to figure out that `.ClientConfig()` applied to `NewNonInteractiveDeferredLoadingClientConfig` and not `&ConfigOverrides`.

**Comment by madhusudancs** [comment]
@nikhiljindal a few very minor nits. LGTM otherwise. Please feel free to apply the LGTM label after addressing those comments.

File: pkg/client/unversioned/clientcmd/client_config.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/client/unversioned/clientcmd/client_config.go]: Done

File: federation/apis/federation/v1alpha1/types.go (hunk 3 above)
Comment (nikhiljindal) [inline federation/apis/federation/v1alpha1/types.go]: Removed

File: federation/apis/federation/v1alpha1/types.go (hunk 2 above)
Comment (nikhiljindal) [inline federation/apis/federation/v1alpha1/types.go]: Done

File: federation/apis/federation/types.go (hunk 1 above)
Comment (nikhiljindal) [inline federation/apis/federation/types.go]: Done. Thanks for catching.

**Comment by nikhiljindal** [comment]
Thanks @madhusudancs for the quick review.
Pushed updated code.

**Comment by madhusudancs** [comment]
@nikhiljindal no problem. Test failures seem real, but I guess you are already working on them. Drawing your attention, just in case.

**Comment by nikhiljindal** [comment]
Yes was fixing the test.
Had to extract out KubeconfigGetter and override it in tests since the default one needs InClusterConfig which expects that the default service account and tokens are setup. It expects env vars like KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT and expects files mounted at `/var/run/secrets/kubernetes.io/serviceaccount/`.
Overrided it by a dummy kubeconfigGetter in test.

Was also exploring if we can do without the deferred client loader. We dont need it since we call loader.ClientConfig() right after creating it. But it requires some refactoring to be able to skip that. Have added a TODO for that.

PTAL.

File: federation/pkg/federation-controller/cluster/cluster_client.go [hunk 5]
```diff
@@ -40,6 +40,32 @@ const (
 	KubeconfigSecretDataKey = "kubeconfig"
 )
 
+// This is to inject a different kubeconfigGetter in tests.
+// We dont use the standard one which calls NewInCluster in tests to avoid having to setup service accounts and mount files with secret tokens.
+var KubeconfigGetterForCluster = func(c *federation_v1alpha1.Cluster) clientcmd.KubeconfigGetter {
```
Comment (madhusudancs) [inline federation/pkg/federation-controller/cluster/cluster_client.go]: It seems somewhat strange that this variable is manipulated/made test compatible by making it a global. It should probably be passed as an argument to `NewClusterClientSet()` or something like that. I understand that this isn't this PR/commit's problem but more about how this package is setup in general. It should probably be refactored in a future PR.

**Comment by madhusudancs** [comment]
Just one comment, but LGTM.

**Comment by mfanjie** [comment]
@nikhiljindal before we test our controller code with unsecured connection, but with this merged PR, it seems the unsecured connection is broken, it always check the namespace and secretRef.
Want to make sure that if we still support unsecured connection, and if there will be a user guideline for setting up the secretRef.

**Comment by nikhiljindal** [comment]
Re: unsecured connection: Yes we should support that. I can update this code to not generate an error if secretRef is empty. Instead, it should just use the default config (without any auth fields set). Will send a PR for this.

Re: user guideline for setting up secretRef: yes we need a user guide for setting up the federation control plane and registering k8s clusters. For setting up secretRef, admins will have to create a secret in the k8s cluster that is hosting the federation control plane. The secret will contain kubeconfig for the corresponding k8s cluster.

Re: checking namespace: Namespace is checked only if explicitly required by admission-control. Default value of --admission-control flag on apiserver is AlwaysAdmit, but we set it to include NamespaceLifecycle at many places. namespace check should go away if you remove NamespaceLifecycle. Ref https://github.com/kubernetes/kubernetes/pull/26298

Apologies for breaking the unsecured connection flow with this PR. Will send a PR to fix it. Am also working on adding an e2e for creating a service.

**Comment by mfanjie** [comment]
btw, why cannot we set the namespace in secretRef but check the environment variable in federation control plane host?

**Comment by nikhiljindal** [comment]
sorry did not understand your comment regarding checking environment variable in federation control plane host. Are you referring to the POD_NAMESPACE env var? What do you want to check?

Re: specifying the namespace in secretRef: cross namespace secretRefs are not allowed i.e pod in namespace A should not be able to access secret in namespace B (this is possible right now, but will not be in future). Hence at all places, secretRef just includes the secret Name and not the namespace (namespace is implied to be the namespace that pod is running in).

Does that answer your question?

**Comment by mfanjie** [comment]
@nikhiljindal yes you answered my question, thanks.

**Pull request merged**